sequence_processing_pipeline/Pipeline.py

```diff
@@ -553,55 +593,62 @@ def generate_sample_info_files(self, addl_info=None):
         """
         if self.mapping_file is not None:
             # Generate a list of BLANKs for each project.
-            df = self.mapping_file[['sample_name', 'project_name']]
+            temp_df = self.mapping_file[[SAMPLE_NAME_KEY, _PROJECT_NAME_KEY]]
+            temp_df_as_dicts_list = temp_df.to_dict(orient='records')
```

Comment:
I think from the use below this could just be an iteration over `temp_df` rather than conversion to `dict` then iteration? To be honest, I'm not sure I understand what's going on below as I would assume the iteration is occurring over `.keys()` since `.items()` isn't used, in which case it's not clear why a `dict` would be used rather than just walking a the `Series`?

Reply:
I think there is something I'm missing? Should this be expressed as `.apply` on the dataframe rather than casting and iterating?

Reply:
The current iteration is not looping over a dictionary--`pandas.DataFrame`'s `to_dict` method is unfortunately named in that, when you pass it the `records` orientation, what it gives you back is a *LIST* (of dictionaries, one per row), not a dictionary itself.  This code is iterating over each row (expressed as a dictionary) and using dictionary's built-in `update` function to add new columns with contents specific to that row.  Since the info to be added arrives from an external function in the form of a dictionary, this meshes well.

Really, I just wrote it this way because it was easy for me.  I have fiddled with it a bit trying to rewrite it to use `apply`, but I have to say that I did not succeed :-/  I am open to education on this!

Reply:
I think it would be something like the below, but it's been a bit since I've done UDFs with pandas, and I'm not sure I really understand all the structures used here... 

Doesn't really matter though, and in writing it I'm not sure if it's necessarily more expressive 

```python
def update_project(row):
    if is_blank(record[SAMPLE_NAME_KEY]):
        proj_info = parse_project_name(row[_PROJECT_NAME_KEY])
        for k, v in proj_info.items():
            row[k] = v
    return row

temp_df = self.mapping_file[[SAMPLE_NAME_KEY, _PROJECT_NAME_KEY]].copy()
temp_df = temp_df.apply(update_project)
```